scalableminds/webknossos
# Upload multiple dataset files

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- if you test on an instance without job support, you can still set `jobsEnabled=true` in `application.conf`. an error toast will appear when wk tries to schedule the conversion.
- open the dataset import view and experiment with various inputs for the upload zone:
  - dragging a zip should analyze whether the zip contains wkw files. if this is not the case, the "Scale" textbox will appear.
  - dragging multiple zips should show a validation error
  - dragging image files should show the "Scale" textbox, too
  - dragging wkw files shouldn't show the scale box
  - mixing wkw and image files should show a validation error
  - dragging tar or rar files should show an error
- try to actually upload some files
  - DnD should support dragging one file, multiple files and folders
  - opening the file picker should allow to select one or multiple files (but no folders)
  - if these (zipped) files are wkw, the provided dataset name should be used to create a new dataset folder in `binaryData`
  - if the files need conversion, `binaryData/sample_organization/.forConversion` should contain a new folder (named after your provided dataset name) which includes the uploaded files
- if more than 10 files are attached to the form, the file list should be compact (only one line per file; possible with a scrolling overflow). otherwise, the second line will show the relative path.
- removing a file via the "remove" action should work

![image](https://user-images.githubusercontent.com/2486553/108863478-8b536a80-75f1-11eb-87e3-0122acc8343f.png)

compact view:
![image](https://user-images.githubusercontent.com/2486553/108863651-b8a01880-75f1-11eb-9b49-14cf7aae7af5.png)


### To Do Backend
- [x] unpack dataset in backend (auto-detect if unzip or move)
- [x] move dataset to destination in backend
- [x] adapt sample dataset import
- [x] adapt cleanUpOrphanFileChunks
- [x] backend: check if all filles have arrived
- [x] error handling (see compiler warnings)
- [x] scala Path vs java Path vs File vs Directory
- [x] ensure no conflict with #5156 (dataset filesize report)
- [x] double-check that files with equal names are not deduplicated if their paths differ
- [x] remove logging

### To Do Frontend
- [x] in resumable upload, pass upload id *and* file id / path
- [x] use complete / progress events instead of fileSuccess / fileProgress
- [x] merge master (especially the changes from #5081
- [x] adapt UI to accept both multiple files and directories (might not work with a single antd element)
- [x] ~defer content checks until upload button is hit ? (files may be dropped one by one)~ I decided to re-validate on every file change. otherwise, one would think that everything is good to go and when hitting "upload" one gets a red warning saying that the scale has to be added. I like it better when one gets instant feedbac.
- [x] show abbreviated file list if too many files
- [X] show relative path in file list? ~unfortunately, this doesn't seem to be possible with antd 3.~ with react-dropzone this works. however, I only added this when fewer than 10 items are present.
- [x] add explanatory text?
- [x] check that progress bar correctly shows progress across all files

### Issues:
- fixes #5033 

------
